pkg/analysis/optionalfields/analyzer.go

```diff
@@ -0,0 +1,540 @@
+	"sigs.k8s.io/kube-api-linter/pkg/analysis/helpers/extractjsontags"
+	"sigs.k8s.io/kube-api-linter/pkg/analysis/helpers/inspector"
+	"sigs.k8s.io/kube-api-linter/pkg/analysis/helpers/markers"
+	"sigs.k8s.io/kube-api-linter/pkg/config"
+
+	"k8s.io/utils/ptr"
+)
+
+const (
+	name = "optionalfields"
+
+	optionalMarker            = "optional"
+	requiredMarker            = "required"
+	kubebuilderOptinalMarker  = "kubebuilder:validation:Optional"
+	kubebuilderRequiredMarker = "kubebuilder:validation:Required"
+
+	minItemsMarker      = "kubebuilder:validation:MinItems"
+	minLengthMarker     = "kubebuilder:validation:MinLength"
+	minPropertiesMarker = "kubebuilder:validation:MinProperties"
+
+	minimumMarker = "kubebuilder:validation:Minimum"
+	maximumMarker = "kubebuilder:validation:Maximum"
+)
+
+func init() {
+	markers.DefaultRegistry().Register(optionalMarker, kubebuilderOptinalMarker)
+}
+
+var (
+	errMarkerMissingValue = errors.New("marker does not have a value")
+)
+
+type analyzer struct {
+	pointerPolicy     config.OptionalFieldsPointerPolicy
+	pointerPreference config.OptionalFieldsPointerPreference
+	omitEmptyPolicy   config.OptionalFieldsOmitEmptyPolicy
+}
+
+// newAnalyzer creates a new analyzer.
+func newAnalyzer(cfg config.OptionalFieldsConfig) *analysis.Analyzer {
+	defaultConfig(&cfg)
+
+	a := &analyzer{
+		pointerPolicy:     cfg.Pointers.Policy,
+		pointerPreference: cfg.Pointers.Preference,
+		omitEmptyPolicy:   cfg.OmitEmpty.Policy,
+	}
+
+	return &analysis.Analyzer{
+		Name: name,
+		Doc: `Checks all optional fields comply with the configured policy.
+		Depending on the configuration, this may include checking for the presence of the omitempty tag or
+		whether the field is a pointer.
+		For structs, this includes checking that if the field is marked as optional, it should be a pointer when it has omitempty.
+		Where structs include required fields, they must be a pointer when they themselves are optional.
+		`,
+		Run:      a.run,
+		Requires: []*analysis.Analyzer{inspector.Analyzer},
+	}
+}
+
+func (a *analyzer) run(pass *analysis.Pass) (any, error) {
+	inspect, ok := pass.ResultOf[inspector.Analyzer].(inspector.Inspector)
+	if !ok {
+		return nil, kalerrors.ErrCouldNotGetInspector
+	}
+
+	inspect.InspectFields(func(field *ast.Field, stack []ast.Node, jsonTagInfo extractjsontags.FieldTagInfo, markersAccess markers.Markers) {
+		a.checkField(pass, field, markersAccess, jsonTagInfo)
+	})
+
+	return nil, nil //nolint:nilnil
+}
+
+func (a *analyzer) checkField(pass *analysis.Pass, field *ast.Field, markersAccess markers.Markers, jsonTags extractjsontags.FieldTagInfo) {
+	if field == nil || len(field.Names) == 0 {
+		return
+	}
+
+	fieldMarkers := markersAccess.FieldMarkers(field)
+
+	fieldName := field.Names[0].Name
+
+	if !isFieldOptional(fieldMarkers) {
+		// The field is not marked optional, so we don't need to check it.
+		return
+	}
+
+	if field.Type == nil {
+		// The field has no type? We can't check if it's a pointer.
```

Comment:
Out of curiosity, what could cause a scenario like this to happen? In this scenario could we still check for the omitempty and skip only the pointer check?